posts/nix-vs-docker-i.md

```diff
@@ -0,0 +1,260 @@
+{ pkgs ? import <nixpkgs> {} }:
+
+pkgs.buildPythonPackage {
+  pname = "myapp";
+  version = "0.1.0";
+  src = ./.;
+  propagatedBuildInputs = [ pkgs.python3Packages.requests ];
+}
+```
+
+No accidental state. No garbage left in `/var`. No need to squash layers. You
+get a single, deduplicated output path like:
+
+```sh
+/nix/store/some-hash-myapp-0.1.0/
+```
+
+That path is input-addressed and reproducible. You can copy it, cache it, or
+reproduce it anywhere. You can even `nix bundle` it to allow running it
+elsewhere, though the API around it is a bit ambiguous and besides the scope of
+today's post.
+
+## Reproducibility
+
+Docker can be made reproducible if you treat it like a hostile system and strip
+away everything non-deterministic. This _is_ possible, but you'll notice very
+quickly that the defaults work against you. Correctness is possible, but it is
+pushed back in the name of 'convenience' for mediocre solutions to difficult
+problems. You can, for example, pin your base image. Vendor your dependencies.
+Disable all dynamic package resolution. Even then, something can—and most likely
+will---still slip through.
+
+In Nix, however, reproducibility is not an extra step. It is part of the system
+design. If your build depends on `openssl`, and the OpenSSL derivation changes,
+the hash of your app changes. If nothing changes, the hash remains the same.
+
+Want to verify the reproducibility of your derivation? Diff it.
+
+```sh
+nix build .#myapp
+nix build .#myapp --rebuild
+diff -r result/ result-2/
+```
+
+Or use `diffoscope`. Regardless, the outputs will be byte-for-byte identical.
+Docker can't really promise that unless you freeze the entire universe.
+
+## Deployment and Distribution
+
+Docker deployments depend on registries. Push your image to Docker Hub, Amazon
+ECR, or GitHub Container Registry. Then pull it down somewhere else. This step
+adds overhead. You're moving entire layered tarballs across the network.
+
+And of course, if you use private registries, you're stuck managing credentials
+and dealing with rate limits or token expiry.
+
+With Nix, you have direct control. You can copy artifacts from one machine to
+another with:
+
+```sh
+nix copy --to ssh://prod-server ./result
+```
+
+Or even to an S3-compatible storage system:
+
+```sh
+nix copy --to s3://my-bucket?region=eu-west-1&endpoint=example.com nixpkgs#hello
+```
+
+No registry required. No daemon [^1] involved. And thanks to the content hash,
+nothing is copied if the store path already exists.
+
+[^1]: Nix _does_ run as a daemon, but unlike the Docker daemon it only runs when
+    you are _building_ things. It is not constantly running, therefore there is
+    a smaller attack vector through the daemon.
+
+If you're working with a public project, using a binary cache like Cachix is
+even simpler. You can build things inside GitHub workflows, upload once, and
+reuse everywhere. And if you really need Docker compatibility?
+
+```nix
+dockerTools.buildImage {
```

Comment:
```suggestion
dockerTools.buildLayeredImage {
```
I think `buildLayeredImage` is pretty much a straight upgrade, no reason not to use it by default. Invalidates the statement about layers below. You can mention how this effectively turns the Docker layer storage into a pseudo-Nix store.